### DISCUSSION TITLE
regarding "n-card game"

### DISCUSSION THREAD

**Discussion by shipingcc** [vuonghy2442/lonelybot#5] [Q&A] [answered]
**Remarkable work!!!** 

I've been following works dedicated on solitaire game for a while, and I'm truly impressed by the capabilities and results of Lonelybot. Your utilization of MCTS for hop solver is particularly intriguing, and I assumed that it would bring groundbreaking results.

I have a few questions about the "n-card game". When I tried the "play"  command, I noticed that the valid DP and DS moves provided resemble the **3-card version** of "K+" (metioned in the article "Searching Solitaire In Real Time" (https://web.engr.oregonstate.edu/~afern/papers/solitaire.pdf)) . The "K+" allows all cards in stock to be moved, trimming out the "draw" move for the 1-card game. 

I wonder if "K+" is the same as the "n-card game" you mentioned in the README. If so, is it convinient to play the "n-card game" with Lonelybot(including all commands)? How can I apply it? Apologies if this is a basic question; I am still getting familiar with Rust.

Thank you for your time, and I look forward to your reply!

**Comment by vuonghy2442** [comment]
Thank you for your interest in my work! 

The n-card game, I mentioned is indeed what you mentioned and I did use the K+ representation from that paper, since my work build up from the Solvitaire paper which also uses the K+ representation. 

Unfortunately, the command line is only limited to 3-card game. If you want to try it immediately, you need to modify the code. Extending it would be quite trivial though (if my code is not buggy). I just overlooked it since I entirely focused on the 3-card games and I haven't tested other cases.

For 1-card game, there would be at least another optimization that can be done for prunning moves (also mentioned in the Solvitaire paper).


For the MCTS stuff, it is not actually as sophisticated as in this paper https://dl.acm.org/doi/10.5555/3037223.3037228 (which i got inspired from). My MCTS stuff is just to optimize where HOP should focus the computation resources on, not wasting time trying the "obviously" bad moves.

I still have some ideas to try on MCTS, but I want to have some time to make sure the current one is actually valid, since it's a bit suspicious that it improves significantly compared to the previous work.

**Reply by vuonghy2442** [reply, marked as answer]
https://github.com/vuonghy2442/lonelybot/blob/dbc55b9cdb4d876d730959d2220490203426847b/lonecli/src/main.rs#L32

This is the line you can modify if you want to try other n-card game :)) sorry, it's a bit inconvenient to rebuild every time

**Reply by vuonghy2442** [reply]
Anyways, I just push a new commit to support other n-card games :)) please try and notify me if there's any bug

**Reply by shipingcc** [reply]
Thank you so much!!!!! I will try it and go explore~
